tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/IbmMq/IbmMqHelper.cs

```diff
@@ -0,0 +1,139 @@
+
+            var span = scope.Span;
+            span.Type = SpanTypes.Queue;
+            span.ResourceName = resourceName;
+            span.SetTag(Tags.SpanKind, SpanKinds.Producer);
+
+            var adapter = new IbmMqHeadersAdapter(message);
+            SpanContextPropagator.Instance.Inject(span.Context, adapter);
+        }
+        catch (Exception ex)
+        {
+            Log.Error(ex, "Error creating or populating scope.");
+        }
+
+        return scope;
+    }
+
+    internal static Scope? CreateConsumerScope(
+        Tracer tracer,
+        CallTargetState state,
+        IMqQueue queue,
+        IMqMessage message)
+    {
+        Scope? scope = null;
+
+        try
+        {
+            var settings = tracer.Settings;
+            if (!settings.IsIntegrationEnabled(IntegrationId.IbmMq))
+            {
+                return null;
+            }
+
+            var parent = tracer.ActiveScope?.Span;
+            var operationName = tracer.CurrentTraceSettings.Schema.Messaging.GetInboundOperationName(MessagingType);
+            if (parent is not null &&
+                parent.OperationName == operationName &&
+                parent.GetTag(Tags.InstrumentationName) != null)
+            {
+                return null;
+            }
+
+            SpanContext? propagatedContext = null;
+            PathwayContext? pathwayContext = null;
+
+            var adapter = new IbmMqHeadersAdapter(message);
+            try
+            {
+                propagatedContext = SpanContextPropagator.Instance.Extract(adapter);
+            }
+            catch (Exception ex)
+            {
+                Log.Error(ex, "Error extracting propagated headers from IbmMq message");
+            }
+
+            var dataStreams = tracer.TracerManager.DataStreamsManager;
+            if (dataStreams.IsEnabled)
+            {
+                try
+                {
+                    pathwayContext = dataStreams.ExtractPathwayContextAsBase64String(adapter);
+                }
+                catch (Exception ex)
+                {
+                    Log.Error(ex, "Error extracting PathwayContext from IbmMq message");
+                }
+            }
+
+            var serviceName = tracer.CurrentTraceSettings.Schema.Messaging.GetServiceName(MessagingType);
+            scope = tracer.StartActiveInternal(
+                operationName,
+                parent: propagatedContext,
+                serviceName: serviceName,
+                finishOnClose: true,
+                startTime: state.StartTime);
+            tracer.TracerManager.Telemetry.IntegrationGeneratedSpan(IntegrationId.IbmMq);
+
+            var resourceName = $"Consume Topic {(string.IsNullOrEmpty(queue.Name) ? "ibmmq" : queue.Name)}";
+
+            var span = scope.Span;
+            span.Type = SpanTypes.Queue;
+            span.ResourceName = resourceName;
+            span.SetTag(Tags.SpanKind, SpanKinds.Consumer);
```

Comment:
I would expect us to be recording more details in the tags here (and likely creating a custom `IbmMqTags` implementation)

Reply:
I've added IbmMqTags with a few basic tags. Overall there's a lot of information we can add, not sure how valuable it is, since I never used IbmMq myself. We can revisit in the future.